homeassistant/components/tplink/switch.py

```diff
@@ -102,46 +103,34 @@ def is_on(self):
         """Return true if switch is on."""
         return self._state
 
-    def turn_on(self, **kwargs):
+    async def async_turn_on(self, **kwargs):
         """Turn the switch on."""
-        self.smartplug.turn_on()
+        await self.smartplug.turn_on()
 
-    def turn_off(self, **kwargs):
+    async def async_turn_off(self, **kwargs):
         """Turn the switch off."""
-        self.smartplug.turn_off()
+        await self.smartplug.turn_off()
 
     @property
     def device_state_attributes(self):
         """Return the state attributes of the device."""
         return self._emeter_params
 
-    @property
-    def _plug_from_context(self):
-        """Return the plug from the context."""
-        children = self.smartplug.sys_info["children"]
-        return next(c for c in children if c["id"] == self.smartplug.context)
-
-    def update(self):
+    async def async_update(self):
         """Update the TP-Link switch's state."""
         try:
+            await self.smartplug.update()
             if not self._sysinfo:
                 self._sysinfo = self.smartplug.sys_info
                 self._mac = self.smartplug.mac
                 self._model = self.smartplug.model
-                if self.smartplug.context is None:
-                    self._alias = self.smartplug.alias
-                    self._device_id = self._mac
-                else:
-                    self._alias = self._plug_from_context["alias"]
-                    self._device_id = self.smartplug.context
-
-            if self.smartplug.context is None:
-                self._state = self.smartplug.state == self.smartplug.SWITCH_STATE_ON
-            else:
-                self._state = self._plug_from_context["state"] == 1
+                self._alias = self.smartplug.alias
```

Comment:
Why set all these as instance variables on the class and not have `name` property return `self.smartplug.alias` ?

Reply:
These are leftovers and are not anymore necessary as the caching is done in the backend. I'll remove the local variables.

One open question here, how to trigger re-read for `device_info` property, in case the device name changes?

Reply:
Call `device_registry.async_update_device` with the new name.

Reply:
Should that be called always under `async_update` (i.e., is there some harm by doing that), or should I restore caching for that specific case?